File: easygallery-core/src/main/java/za/co/bwmuller/easygallerycore/data/loader/AlbumLoader.java
package za.co.bwmuller.easygallerycore.data.loader;

import android.content.Context;
import android.database.Cursor;
import android.os.Bundle;
import androidx.fragment.app.FragmentActivity;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.Loader;

import java.lang.ref.WeakReference;

import za.co.bwmuller.easygallerycore.data.cursor.AlbumCursor;

/**
 * Created by Bernhard Müller on 8/23/2017.
 */

public class AlbumLoader implements LoaderManager.LoaderCallbacks<Cursor> {
    private WeakReference<Context> mContext;
    private LoaderManager mLoaderManager;
    private AlbumCallback mAlbumCallback;

    @Override public Loader<Cursor> onCreateLoader(int id, Bundle args) {
        Context context = mContext.get();
        if (context == null) {
            return null;
        }
        return AlbumCursor.newInstance(context, mAlbumCallback.getConfig());
    }

    @Override public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
        Context context = mContext.get();
        if (context == null) {
            return;
        }

        mAlbumCallback.onCursorLoad(data);
    }

    @Override public void onLoaderReset(Loader<Cursor> loader) {
        Context context = mContext.get();
        if (context == null) {
            return;
        }

        mAlbumCallback.onCursorReset();
    }

    public void onCreate(FragmentActivity activity, AlbumCallback albumCallback) {
        mContext = new WeakReference<Context>(activity);
        mAlbumCallback = albumCallback;
        mLoaderManager = activity.getSupportLoaderManager();
    }

    public void onDestroy() {
        mLoaderManager.destroyLoader(getLoaderId());
        mAlbumCallback = null;
    }

    private int getLoaderId() {
        return AlbumLoader.class.getSimpleName().hashCode() + mAlbumCallback.getConfig().loaderScope.ordinal();
    }

    public void loadAlbums() {
        mLoaderManager.initLoader(getLoaderId(), null, this);
    }

//    public int getCurrentSelection() {
//        return mCurrentSelection;
//    }
//
//    public void setStateCurrentSelection(int currentSelection) {
//        mCurrentSelection = currentSelection;
//    }
}

File: easygallery-core/src/main/java/za/co/bwmuller/easygallerycore/data/loader/AlbumLoader.kt
package za.co.bwmuller.easygallerycore.data.loader

import android.content.Context
import android.database.Cursor
import android.os.Bundle
import androidx.fragment.app.FragmentActivity
import androidx.loader.app.LoaderManager
import androidx.loader.app.LoaderManager.LoaderCallbacks
import androidx.loader.content.Loader
import za.co.bwmuller.easygallerycore.data.cursor.AlbumCursor.Companion.newInstance
import java.lang.ref.WeakReference

/**
 * Created by Bernhard Müller on 8/23/2017.
 */
class AlbumLoader(activity: FragmentActivity, albumCallback: AlbumCallback?) : LoaderCallbacks<Cursor> {

    private var mContext: WeakReference<Context> = WeakReference(activity)
    private var mLoaderManager: LoaderManager = LoaderManager.getInstance(activity)
    private var mAlbumCallback: AlbumCallback? = albumCallback

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<Cursor> {
        return newInstance(mContext.get()!!, mAlbumCallback?.config!!)
    }

    override fun onLoadFinished(loader: Loader<Cursor>, data: Cursor) {
        val context = mContext.get() ?: return
        mAlbumCallback?.onCursorLoad(data)
    }

    override fun onLoaderReset(loader: Loader<Cursor>) {
        val context = mContext.get() ?: return
        mAlbumCallback?.onCursorReset()
    }

    fun onDestroy() {
        mLoaderManager.destroyLoader(loaderId)
        mAlbumCallback = null
    }

    private val loaderId: Int
        get() = AlbumLoader::class.java.simpleName.hashCode() + (mAlbumCallback?.config?.loaderScope?.ordinal ?: 0)

    fun loadAlbums() {
        mLoaderManager.initLoader(loaderId, null, this)
    }
}